### ISSUE TITLE
Dogwrap doesn't support unicode characters

### ISSUE THREAD

**Issue by synth** [DataDog/datadogpy#371]
This is a follow up to #223 

There are two issues at play here:
1. Wrap.py doesn't handle subprocess output with unicode properly and causes fatal exception.
2. Wrap.py should be robust enough to handle any exception without causing fatal error so as to not cause subprocess to fail. 

I will go into details on each. 

This is in reference to dogwrap v0.28.0 on Docker Linux. The issue does not occur on Mac OSX 10.14.4. I have not tested on other OS's.
 
### Wrap.py handling of unicode characters
Test case (echo_unicode.sh):
```
echo "This is one string: øø"
echo 'A unicode \xc6\x8e string \xc3\xb1'
```

Command:
```
dogwrap -k $API_KEY -n 'test unicode' --submit_mode all './echo_unicode'
```

Exception:
```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/dist-packages/datadog/dogshell/wrap.py", line 69, in run
    self._fwd_out.write(line.decode(fwd_out_encoding))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 20: ordinal not in range(128)
```

### Wrap.py handling of exceptions

The dogwrap python script fails completely when there is a unicode character output as described above. This is extremely problematic because it causes subprocess to fail. In our case, the subprocess is DelayedJob (DJ) which seems to frequently work with database transactions. When dogwrap fails, it causes the DJ process to be killed leaving a database transaction open which causes all other DJ processes to hang wreaking utter havoc on our database .

The solution should be more robust exception handling to not let the subprocess fail - after all this is just a wrapper script. 
 
You can see in the python code that the error handling needs to be improved. The error occurs in OutputReader which is itself wrapped in a try/except block, but the except block only catches Timeout exceptions. Recommend to add additional catch on all Exception types:

https://github.com/DataDog/datadogpy/blob/bc3321558941ff04ccbfac3b8e26268b6942d4f3/datadog/dogshell/wrap.py#L112-L152

**Comment by nmuesch** [comment]
Hey @synth Thanks for the detailed report with reproduction steps. I was taking a look into this one and it looks like the issue is actually related to the encoding format tied to stdout/stderr. I wasn't able to reproduce this issue on machines that had the filesystem encoding set to 'UTF-8', however I was able to reproduce when I overrode those to 'ascii' instead. 

I believe this is the reason that you weren't able to reproduce on Mac OSX but were able to on Linux. Would it be possible to set the environment variable `PYTHONIOENCODING` to `UTF-8` as a workaround here? More information about this environment variable is here - https://docs.python.org/3/using/cmdline.html#envvar-PYTHONIOENCODING, or set your locale to UTF-8 if it makes sense here. 

It looks like the issue is actually when trying to write the unicode output back to standard output here - https://github.com/DataDog/datadogpy/blob/master/datadog/dogshell/wrap.py#L69 It seems like the process being run has output in an encoding that doesn't match the locale of the filesystem.

**Comment by synth** [comment]
Hi @nmuesch thanks for your investigation. We are using standard Ruby docker image. Our ruby is likely UTF-8, I'm not sure what standard Ruby docker locale is. From the look of it, it shows POSIX: 

```
[root@ip-10-10-10-145 /]# docker exec -it dc6f5fb12bb4 locale
LANG=
LANGUAGE=
LC_CTYPE="POSIX"
LC_NUMERIC="POSIX"
LC_TIME="POSIX"
LC_COLLATE="POSIX"
LC_MONETARY="POSIX"
LC_MESSAGES="POSIX"
LC_PAPER="POSIX"
LC_NAME="POSIX"
LC_ADDRESS="POSIX"
LC_TELEPHONE="POSIX"
LC_MEASUREMENT="POSIX"
LC_IDENTIFICATION="POSIX"
LC_ALL=
```

That said, shouldn't this script:
1. Be resilient against situations where the encoding a script doesn't match the file system
2. Be resilient against exceptions of any kind from killing the subprocess without SIGTERM?

**Comment by nmuesch** [comment]
Thanks for the info. I'm not overly familiar with all the available options, but I imagine the local of the Mac you attempted to reproduce on all had `en_US.utf8` values? 

> Be resilient against situations where the encoding a script doesn't match the file system
I think the primary issue is that the output of the script that you pass you dogwrap contains characters that don't match the supported character set of the filesystem.

I think it's a tricky issue. Ultimately the children threads are what listen to and collect the stdout/err from your script. However, we only catch exceptions from the main thread. So if the children threads have a problem reading any of the scripts output, the child thread throws an exception and the parent thread (dogwrap) continues to emit the event with what it has and doesn't attempt to kill the script that was triggered. One possible solution, if you'd be ok with that, would be to ignore errors and simply not print the characters that can't be encoded/decoded, that way we're more resilient to this exception going forward.

For my understanding, you say you have a scenario where your script is killed because of this exception? In my testing and read of the code it seems like the script's execution should continue until it naturally finishes, despite dogwrap not fully being able to parse the output. Do you happen to have any more context about why your script is failing in this case? 

> Be resilient against exceptions of any kind from killing the subprocess without SIGTERM?
I don't believe we are hitting a scenario where dogwrap is actually killing the subprocess here. I believe that it would be connected to my note above but happy to continue looking if this is indeed the case.

**Comment by synth** [comment]
Hi @nmuesch,

So, I wrote a test to check if a subprocess gets killed with the parent, and according to my test case, it does not as you have indicated. For reference, here is my test code (amendment of `echo_unicode` script - called in same way as above): 

```
echo "Starting to sleep" | tee -a echo_log
sleep 5
echo "This is one string: øø" | tee -a echo_log
sleep 10
echo "Done" | tee -a echo_log
```

The final command does get appended to the log even though dogwrap has failed. That said, we are still encountering the situation that at the exact time when dogwrap script fails, we have orphaned database transactions which I can only explain by the DelayedJob process exiting without SIGTERM. 

Unfortunately, my python knowledge and Linux subprocess knowledge are hitting their limits. 

At the moment, we are testing changing the locale of our Docker container and will report back.

**Comment by nmuesch** [comment]
Thanks for this and your testing as well! Let me know how the docker locale change goes. 

What would you think of this for being more resilient here - 

> One possible solution, if you'd be ok with that, would be to ignore errors and simply not print the characters that can't be encoded/decoded, that way we're more resilient to this exception going forward.

**Comment by synth** [comment]
@nmuesch Yes, I think that sounds good.

**Comment by synth** [comment]
@nmuesch Doesn't look like changing locale works:
```
root@recognize-rails:/usr/src/app# dogwrap -k $DataDogAPIKey -n 'test unicode' --submit_mode all './echo_unicode.sh'
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/dist-packages/datadog/dogshell/wrap.py", line 69, in run
    self._fwd_out.write(line.decode(fwd_out_encoding))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 20: ordinal not in range(128)

root@recognize-rails:/usr/src/app# locale
locale: Cannot set LC_CTYPE to default locale: No such file or directory
locale: Cannot set LC_MESSAGES to default locale: No such file or directory
locale: Cannot set LC_ALL to default locale: No such file or directory
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

**Comment by ofek** [comment]
@synth In my Dockerfiles I usually do:

```
# Ensure proper locale
RUN localedef -i en_US -c -f UTF-8 -A /usr/share/locale/locale.alias en_US.UTF-8
ENV LANG=en_US.utf8
ENV LANGUAGE=en_US.utf8
ENV LC_ALL=en_US.utf8
```

Does that work?

**Comment by nmuesch** [comment]
I just merged in #372 which should allow dogwrap to be more resilient to this error.

**Comment by synth** [comment]
Unfortunately, @nmuesch @ofek, neither of the two approaches above fixed the issue:

<img width="1010" alt="Screenshot 2019-04-29 17 29 53" src="https://user-images.githubusercontent.com/10546/56935214-82042000-6aa4-11e9-8497-607bbeb02d01.png">

You can see that the "ignore" is referenced above which confirms we're using the latest datadogpy version.

**Comment by synth** [comment]
So, some more context is that this issue is present on the AWS ECS (elastic container) service which runs our docker workloads. When I run the cron's command manually on a docker container, there is no issue. However, when launched by AWS service it crashes. 

Command line invocation:
<img width="1675" alt="Screenshot 2019-04-30 01 01 47" src="https://user-images.githubusercontent.com/10546/57001300-c3f89900-6b6c-11e9-89fa-99ce15c94656.png">

AWS invocation of same exact command
<img width="1461" alt="Screenshot 2019-04-30 01 09 13" src="https://user-images.githubusercontent.com/10546/57001305-ca871080-6b6c-11e9-9673-f04d7e748cd8.png">
 
Please note that in the above cases, I'm using a forked version of datadogpy that wraps the line at issue with try/except block. See my changes here: https://github.com/DataDog/datadogpy/compare/master...Recognize:synth/fix-unicode-output?expand=1

We still don't know what's causing this but we feel Datadog should implement similar exception handling to not crash and cause this catastrophic failure we are seeing. Otherwise, this is a complete blocker to Datadog adoption for monitoring of our infrastructure.

**Comment by nmuesch** [comment]
Hey @synth Sorry to hear your still having this issue, but thank you for the additional information and follow up 🙇 Just wanted to let you know I've re opened this issue and we'll take a look shortly.

**Comment by synth** [comment]
Thank you! This is such a strange one!

**Comment by synth** [comment]
Hi @nmuesch We are also hard at work trying to make sense of this issue. In our testing, we found another related but not quite exact issue. With the following command: 

`dogwrap -n 'Test' -k $DataDogApiKey --submit_mode all "echo héllø"`

We get: 
```
Failed to execute ['echo', 'h\xc3\xa9ll\xc3\xb8']
Traceback (most recent call last):
  File "/usr/local/bin/dogwrap", line 11, in <module>
    load_entry_point('datadog==0.28.0', 'console_scripts', 'dogwrap')()
  File "/usr/local/lib/python2.7/dist-packages/datadog/dogshell/wrap.py", line 274, in main
    options.proc_poll_interval, options.buffer_outs)
  File "/usr/local/lib/python2.7/dist-packages/datadog/dogshell/wrap.py", line 108, in execute
    proc = subprocess.Popen(u' '.join(cmd), stdout=subprocess.PIPE,
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 1: ordinal not in range(128)
```

This time you can see its failing in the Popen code rather than the OutputReader.

**Comment by nmuesch** [comment]
Hey @synth I can't seem to reproduce the issue on the `ruby:2.6` image with the #372 fix in place and the command you mentioned. Could you share the image tag you're using and I'll try to reproduce again on my side?

**Comment by synth** [comment]
@nmuesch I'm having trouble creating a perfect test case for you to reproduce due to some weird issues with this script. It seems like this script tokenizes the command and processes in a strange way. We see some parts of the wrapped command run before others. We work around it by wrapping a whole command in a `bash -c "..."` call. Do you know why this is? We can see the tokenization in the failure output with the following (imperfect) test case

1. Create a file called `test_dd_unicode` with:
```
var1="Mich\xc3\xa9lle"
var2="helløøééé"
echo "Running Test DD Unicode Script"
apt-get update -qq && apt-get install -y build-essential libpq-dev nodejs python-pip locales && pip install git+https://github.com/datadog/datadogpy.git@master

sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && locale-gen
export LANG=en_US.UTF-8 
export LANGUAGE=en_US:en  
export LC_ALL=en_US.UTF-8 

locale
echo ENV LANG
echo LANGUAGE
echo LC_ALL

echo $var1 $var2

dogwrap -n 'TestDDUnicode' -k $DataDogAPIKey --submit_mode all "echo 'Starting Unicode Test' && bash -c 'ruby -e \"puts $var1, $var2 \"' && echo 'Completed DD Unicode Test'"
echo "Finished Test DD Unicode Script"
```
2. Then run via Docker with: 
```
docker run -it --rm --name TestDDUnicode -e DataDogApiKey=$DataDogApiKey -v "$PWD":/usr/src/myapp -w /usr/src/myapp ruby:2.6 /bin/bash ./test_dd_unicode
```

I tested with ruby image for 2.5.1 (the one we're using) and 2.6 - both experience the issue  which give this output:

```
Mich\xc3\xa9lle helløøééé
Failed to execute ['all', 'echo', "'Starting", 'Unicode', "Test'", '&&', 'bash', '-c', "'ruby", '-e', '"puts', 'Mich\\xc3\\xa9lle,', 'hell\xc3\xb8\xc3\xb8\xc3\xa9\xc3\xa9\xc3\xa9', '"\'', '&&', 'echo', "'Completed", 'DD', 'Unicode', "Test'"]
Traceback (most recent call last):
  File "/usr/local/bin/dogwrap", line 11, in <module>
    load_entry_point('datadog==0.28.0', 'console_scripts', 'dogwrap')()
  File "/usr/local/lib/python2.7/dist-packages/datadog/dogshell/wrap.py", line 275, in main
    options.proc_poll_interval, options.buffer_outs)
  File "/usr/local/lib/python2.7/dist-packages/datadog/dogshell/wrap.py", line 107, in execute
    proc = subprocess.Popen(u' '.join(cmd), stdout=subprocess.PIPE,
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 4: ordinal not in range(128)
Finished Test DD Unicode Script
```

So a couple of things about the above:
1. The command is using variables that come from bash script which results in encoding failure in a different place than what we experience in our app. Here it fails in the `subprocess.Popen` call, rather then the OutputReader. 
2. You can see the tokenization in the "Failed to execute..." line.

**Comment by synth** [comment]
Also, I've created a 2nd test case that more closely reproduces the issue but I can't seem to get it to run correctly. This case involves running a ruby script that pulls unicode data from a remote source (ie, a gist) and outputs it. I've put this in a repo you can clone: 

https://github.com/synth/dd-unicode-test

**Comment by synth** [comment]
Ok, so, I've unblocked myself on the 2nd command error by modifying wrap.py with:

https://github.com/Recognize/datadogpy/blob/c8c327a757ef9a425187dbb3706f0b07fcf47734/datadog/dogshell/wrap.py#L266-L271

    for part in args:
        part = unicode(part, 'utf-8')
        cmd.extend(part.split(' '))
        
    if cmd[0] == "all":
        cmd.pop(0)

+ Adding the unicode line to the arg parts seems like a good idea across the board
+ I'm not sure why in my test app, the "all" flag shows up in the command, but don't see this in our production environment. This is obviously a hack.

So, with the above, I'm actually able to successfully run my test commands and get unicode output with echo, ruby, and even rails. So, I'm scratching my head. 

My next stab at this will be to amend the test app to more closely reproduce our environment with using a database and delayed job as those use separate log streams and maybe there is an issue with how those output.

**Comment by synth** [comment]
@nmuesch do you have any updates here? We haven't been able to give more time to this and are hoping DD can pick up on this given the information above and the severity of the issue. We are still using a forked version to avoid the critical issue brought on by this bug :(

**Comment by zippolyte** [comment]
Hello @synth, 

I've opened this PR https://github.com/DataDog/datadogpy/pull/379 to try and solve the issues you are facing with dogwrap. I've successfully managed to run the command in your test repo https://github.com/synth/dd-unicode-test (thanks a lot for that). 
Are you able to run it on your environment as well ?

As for why you had problems with the `all` flag:
> I'm not sure why in my test app, the "all" flag shows up in the command, but don't see this in our production environment

This is because the `$DataDogAPIKey` environment variable is not set (you set `DataDogApiKey`, notice capitalization), so it messes up the argument parsing, and the script thinks `all` is part of the command.
